contracts/zapper-contract/SnowglobeZapperVector.sol

```diff
@@ -0,0 +1,140 @@
+
+        address token0 = pair.token0();         // first token in the pool
+        address token1 = pair.token1();         // second token in the pool 
+
+        bool isInputA = token0 == tokenIn;
+        require(isInputA || token1 == tokenIn, "Input token not present in liquidity pair");
+
+        address[] memory path = new address[](2);
+        path[0] = tokenIn;
+        path[1] = isInputA ? token1 : token0;
+
+        uint256 fullInvestment = IERC20(tokenIn).balanceOf(address(this));
+        uint256 swapAmountIn;
+        if (isInputA) {
+            swapAmountIn = _getSwapAmount(fullInvestment, reserveA, reserveB);
+        } else {
+            swapAmountIn = _getSwapAmount(fullInvestment, reserveB, reserveA);
+        }
+
+        _approveTokenIfNeeded(path[0], address(router));
+         uint256[] memory swappedAmounts = IJoeRouter(router)
+            .swapExactTokensForTokens(
+                swapAmountIn,
+                tokenAmountOutMin,
+                path,
+                address(this),
+                block.timestamp
+            );
+
+        _approveTokenIfNeeded(path[1], address(router));
+        (, , uint256 amountLiquidity) = IJoeRouter(router).addLiquidity(
+            path[0],
+            path[1],
+            fullInvestment.sub(swappedAmounts[0]),
+            swappedAmounts[1],
+            1,
+            1,
+            address(this),
+            block.timestamp
+        );
+
+        _approveTokenIfNeeded(address(pair), address(vault));
+        vault.deposit(amountLiquidity);
+
+        //taking receipt token and sending back to user
+        vault.safeTransfer(msg.sender, vault.balanceOf(address(this)));
+
+        //interact with gauge proxy to get gauge address
+        address gaugeAddress = IGaugeProxyV2(gaugeProxy).getGauge(snowglobe);
+
+        //deposit for into gauge
+        IGaugeV2(gaugeAddress).depositFor(vault.balanceOf(msg.sender), msg.sender);
+
+        _returnAssets(path);
+    }
+
+    function _getSwapAmount(uint256 investmentA, uint256 reserveA, uint256 reserveB) public view override returns (uint256 swapAmount) {
+        uint256 halfInvestment = investmentA.div(2);
+        uint256 nominator = IJoeRouter(router).getAmountOut(
+            halfInvestment,
+            reserveA,
+            reserveB
+        );
+        uint256 denominator = IJoeRouter(router).quote(
+            halfInvestment,
+            reserveA.add(halfInvestment),
+            reserveB.sub(nominator)
+        );
+        swapAmount = investmentA.sub(
+            Babylonian.sqrt(
+                (halfInvestment * halfInvestment * nominator) / denominator
+            )
+        );
+    }
+
+    function estimateSwap(address snowglobe, address tokenIn, uint256 fullInvestmentIn) public view returns (uint256 swapAmountIn, uint256 swapAmountOut, address swapTokenOut){
+        (, IUniPair pair) = _getVaultPair(snowglobe);
```

Comment:
We are using the universal pair here, but the Joe Router in line 133? Why are we using the Joe Router?

Reply:
traderjoe is the only one with liquidity for xptp-ptp swap